File: cdk_lib/api_gw.py
from typing import Tuple

from aws_cdk import aws_apigateway, aws_iam, aws_s3, core


class RestApi(aws_apigateway.RestApi):
    def __init__(self, scope: core.Construct, construct_id: str, **kwargs) -> None:
        defaults = {
            'endpoint_configuration': aws_apigateway.EndpointConfiguration(types=[aws_apigateway.EndpointType.REGIONAL])
        }
        super().__init__(scope, construct_id, **defaults, **kwargs)

    def add_s3_integration(self, path: str, *,
                           bucket: aws_s3.Bucket,
                           s3_key_prefix: str = None,
                           http_method: str = 'POST',
                           s3_key_suffix: str = '-request.json',
                           role: aws_iam.Role = None,
                           ) -> None:
        if not role:
            role = self.add_s3_integration_role(bucket)
        if not s3_key_prefix:
            s3_key_prefix = path

        s3_integration, s3_integration_options = self.s3_aws_integration(bucket.bucket_name, s3_key_prefix, role)
        self.root.resource_for_path(path).add_method(http_method, s3_integration,
                                                     method_responses=[
                                                         aws_apigateway.MethodResponse(status_code='202')],
                                                     request_parameters={
                                                         'method.request.header.Content-Type': False,
                                                         'method.request.header.b3': False,
                                                         'method.request.header.x-b3-traceid': False,
                                                         'method.request.header.x-b3-spanid': False,
                                                         'method.request.header.x-b3-sampled': False,
                                                         'method.request.header.x-b3-flags': False,
                                                     })
        self.latest_deployment.add_to_logical_id({
            'path': f'{bucket.bucket_name}/{s3_key_prefix}/{{requestId}}{s3_key_suffix}',
            'integration_responses': s3_integration_options.integration_responses,
            'request_parameters': s3_integration_options.request_parameters,
        })

    @staticmethod
    def s3_aws_integration(bucket_name: str, s3_key_prefix: str, role: aws_iam.Role
                           ) -> Tuple[aws_apigateway.AwsIntegration, aws_apigateway.IntegrationOptions]:
        integration_options = aws_apigateway.IntegrationOptions(
            credentials_role=role,
            integration_responses=[aws_apigateway.IntegrationResponse(
                status_code='202',
                selection_pattern='2\\d{2}',
                response_templates={'application/json': '{"requestId": "$context.requestId"}'}
            )],
            passthrough_behavior=aws_apigateway.PassthroughBehavior.WHEN_NO_MATCH,
            request_parameters={
                'integration.request.header.Content-Type': 'method.request.header.Content-Type',
                # we can save headers as as meta data on the S3 object
                'integration.request.header.x-amz-meta-b3': 'method.request.header.b3',
                'integration.request.header.x-amz-meta-x-b3-traceid': 'method.request.header.x-b3-traceid',
                'integration.request.header.x-amz-meta-x-b3-spanid': 'method.request.header.x-b3-spanid',
                'integration.request.header.x-amz-meta-x-b3-sampled': 'method.request.header.x-b3-sampled',
                'integration.request.header.x-amz-meta-x-b3-flags': 'method.request.header.x-b3-flags',
                'integration.request.path.requestId': 'context.requestId',
            }
        )
        aws_integration = aws_apigateway.AwsIntegration(service='s3',
                                                        integration_http_method="PUT",
                                                        path=f'{bucket_name}/{s3_key_prefix}/{{requestId}}-request.json',
                                                        options=integration_options)
        return aws_integration, integration_options

    def add_s3_integration_role(self, bucket: aws_s3.Bucket) -> aws_iam.Role:
        role = aws_iam.Role(self, f'{bucket.node.id}ApigwRole',
                            assumed_by=aws_iam.ServicePrincipal('apigateway.amazonaws.com').grant_principal)
        bucket.grant_read_write(role)
        return role

File: cdk/lib/common/api-gw.ts
import {AwsIntegration, EndpointType, IntegrationOptions, PassthroughBehavior, RestApi} from "@aws-cdk/aws-apigateway"
import {Bucket} from "@aws-cdk/aws-s3"
import {Role, ServicePrincipal} from "@aws-cdk/aws-iam"
import {RestApiProps} from "@aws-cdk/aws-apigateway/lib/restapi"
import {Construct} from "@aws-cdk/core"


export interface S3IntegrationProps {
    bucket: Bucket,
    s3KeyPrefix?: string,
    s3KeySuffix?: string,
    httpMethod?: string,
    role?: Role
}

export class RestApiWithDefaults extends RestApi {
    constructor(scope: Construct, construct_id: string, props?: RestApiProps) {
        super(scope, construct_id, {endpointConfiguration: {types: [EndpointType.REGIONAL], ...props}})
    }

    addS3Integration(path: string, props: S3IntegrationProps): void {
        const role = props.role || this.addS3IntegrationRole(props.bucket)
        const s3KeyPrefix = props.s3KeyPrefix || path
        const s3KeySuffix = props.s3KeyPrefix || '-request.json'
        const httpMethod = props.httpMethod || 'POST'

        const {
            s3AwsIntegration,
            integrationOptions
        } = this.createS3AwsIntegration(props.bucket.bucketName, s3KeyPrefix, s3KeySuffix, role)
        this.root.resourceForPath(path).addMethod(httpMethod, s3AwsIntegration, {
            methodResponses: [{statusCode: '202'}],
            requestParameters: {
                'method.request.header.Content-Type': false,
                'method.request.header.b3': false,
                'method.request.header.x-b3-traceid': false,
                'method.request.header.x-b3-spanid': false,
                'method.request.header.x-b3-sampled': false,
                'method.request.header.x-b3-flags': false
            }
        })
        this.latestDeployment?.addToLogicalId({
            path: `${props.bucket.bucketName}/${s3KeyPrefix}/{requestId}${s3KeySuffix}`,
            'integration_responses': integrationOptions.integrationResponses,
            'request_parameters': integrationOptions.requestParameters,
        })
    }

    createS3AwsIntegration(bucketName: string, s3KeyPrefix: string, s3KeySuffix: string, role: Role)
        : { s3AwsIntegration: AwsIntegration, integrationOptions: IntegrationOptions } {
        const integrationOptions: IntegrationOptions = {
            credentialsRole: role,
            integrationResponses: [{
                statusCode: '202',
                selectionPattern: '2\\d{2}',
                responseTemplates: {'application/json': '{"requestId": "$context.requestId"}'}
            }],
            passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
            requestParameters: {
                'integration.request.header.Content-Type': 'method.request.header.Content-Type',
                // we can save headers as as meta data on the S3 object
                'integration.request.header.x-amz-meta-b3': 'method.request.header.b3',
                'integration.request.header.x-amz-meta-x-b3-traceid': 'method.request.header.x-b3-traceid',
                'integration.request.header.x-amz-meta-x-b3-spanid': 'method.request.header.x-b3-spanid',
                'integration.request.header.x-amz-meta-x-b3-sampled': 'method.request.header.x-b3-sampled',
                'integration.request.header.x-amz-meta-x-b3-flags': 'method.request.header.x-b3-flags',
                'integration.request.path.requestId': 'context.requestId',
            }
        }
        const s3AwsIntegration = new AwsIntegration({
            service: 's3',
            integrationHttpMethod: 'PUT',
            path: `${bucketName}/${s3KeyPrefix}/{requestId}${s3KeySuffix}`,
            options: integrationOptions
        })
        return {s3AwsIntegration, integrationOptions}
    }

    addS3IntegrationRole(bucket: Bucket): Role {
        const role = new Role(this, `${bucket.node.id}ApigwRole`, {
            assumedBy: new ServicePrincipal('apigateway.amazonaws.com').grantPrincipal
        })
        bucket.grantReadWrite(role)
        return role
    }
}
